http3/dispatch/connection_pool.py

```diff
@@ -110,21 +110,35 @@ def num_connections(self) -> int:
         cert: CertTypes = None,
         timeout: TimeoutTypes = None,
     ) -> AsyncResponse:
-        connection = await self.acquire_connection(request.url.origin)
-        try:
-            response = await connection.send(
-                request, verify=verify, cert=cert, timeout=timeout
+        allow_connection_reuse = True
+        connection = None
+        while connection is None:
+            connection = await self.acquire_connection(
+                origin=request.url.origin, allow_connection_reuse=allow_connection_reuse
             )
-        except BaseException as exc:
-            self.active_connections.remove(connection)
-            self.max_connections.release()
-            raise exc
+            try:
+                response = await connection.send(
+                    request, verify=verify, cert=cert, timeout=timeout
+                )
+            except BaseException as exc:
+                self.active_connections.remove(connection)
+                self.max_connections.release()
+                if isinstance(exc, NotConnected) and allow_connection_reuse:
+                    connection = None
+                    allow_connection_reuse = False
+                else:
+                    raise exc
```

Comment:
I'm not sure if you meant to remove the full traceback here but currently that's what occurs. May want to call just `raise`?

Reply:
I think it's only if it was `raise exc from None` that it'd remove the full traceback, right?

Ultimately we'll want to specifically catch all connection error types that we expect, and re-raise them as specific `http3` exception types, so that we can have a nice narrowly-defined set of exceptions that the user can expect to see.

Will need to do some fairly compresensive review once we start narrowing in on a 1.0 in order to be confident that've got a clear exception contract in place around the API.

Reply:
Definitely want to handle all non-library exceptions before they go to users.

I'm pretty sure that reraising the way we're doing does recreate the traceback:

```python
def f():
    try:
        g()
    except RuntimeError as e:
        raise e

def g():
    raise RuntimeError("!")

f()
```

gives:

```
Traceback (most recent call last):
  File "/tmp/r.py", line 11, in <module>
    f()
  File "/tmp/r.py", line 5, in f
    raise e
RuntimeError: !
```

and

```python
def f():
    try:
        g()
    except RuntimeError as e:
        raise 

def g():
    raise RuntimeError("!")

f()
```

gives:

```
Traceback (most recent call last):
  File "/tmp/r.py", line 11, in <module>
    f()
  File "/tmp/r.py", line 3, in f
    g()
  File "/tmp/r.py", line 9, in g
    raise RuntimeError("!")
RuntimeError: !
```